### PR TITLE
CU-ev2g8f Fix Sequence#travserse, Option#apEval & Either#apEval

### PR DESCRIPTION
**Pull request by nomisRev**
This PR fixes two issues for traversing an `Sequence`.

- Sequence#traverse was implemented using `foldRight` without `Eval` which first materialises the whole list in memory. This _always_ results in an `OOM` even if the transformation of `traverse` results in a _short-circuit_.
- 
- Fix `apEval` for `Either` and `Option` to properly _short-circuit_.

### CODE REVIEW

**Comment by franciscodr** [comment]
Task linked: [CU-ev2g8f Fix Sequence#traverse](https://app.clickup.com/t/ev2g8f)

File: arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt [hunk 1]
```diff
@@ -1588,7 +1588,7 @@ fun <A, B, C> EitherOf<A, B>.ap(ff: EitherOf<A, (B) -> C>): Either<A, C> =
   flatMap { a -> ff.fix().map { f -> f(a) } }
 
 fun <A, B, C> Either<A, B>.apEval(ff: Eval<Either<A, (B) -> C>>): Eval<Either<A, C>> =
-  ff.map { this.ap(it) }
+  fold({ l -> Eval.now(l.left()) }, { r -> ff.map { it.map { f -> f(r) } } })
```
Comment (aballano) [inline arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt:1591]: is there or should be a specific test for this change?

**Review by aballano — APPROVED** [review]

**Review by i-walker — APPROVED** [review]
Thank you

**Review by raulraja — APPROVED** [review]
thanks @nomisRev

File: arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt (hunk 1 above)
Comment (nomisRev) [reply arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt:1591]: We should add many tests.. 😭😅 This code was taken from the `EitherApply.kt` code, and is verified in `TraverseTest`. By reverting this change it makes `TraverseTest` fail.

Comment (aballano) [reply arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt:1591]: > By reverting this change it makes TraverseTest fail.

That's what I meant, perfect then! 👌

**Pull request merged**